### DISCUSSION TITLE
--reload-exclude cannot recursively ignore directories with '**' glob; changes outside --reload-dir trigger reload

### DISCUSSION THREAD

**Discussion by AntiSol** [Kludex/uvicorn#2526] [Potential Issue]
Using uvicorn 0.32.1 (installed via pip) and python 3.9.18, and the attached directory structure (zipfile: [uvicorn_test.zip](https://github.com/user-attachments/files/18003032/uvicorn_test.zip)), it seems that uvicorn is:
a) not properly honouring globs including the '**' pattern and 
b) not properly honouring --reload-dir

steps to reproduce:
1. extract zipfile into an empty dir, cd to that dir so that pwd is the root of the zip/project
2. `python3 -m venv venv`
3. `. venv/bin/activate`
4. `pip install -r requirements.txt`
5. `uvicorn --host 0.0.0.0 --port 4242 --reload --reload-dir myapp --reload-dir mylib/ --reload-exclude 'infra/**' myapp.api.main:app`
6. modify a file in a subdirectory of infra infra/ e.g with `echo "changed" >> infra/packages/foo/bar/baz.py`
7. observe issue: uvicorn reloads despite 
    a) --reload-dir specifying only the 2 directories and 
    b) --reload-exclude specifically excluding subdirectories of infra/ using the '**' glob pattern.
8. observe that chanegs to infra/a_file.py do _not_ cause a reload - it seems that --reload-exclude works, but only one level deep.
9. Note that using --reload-exclude 'infra/**' causes a substantial delay at startup if there are many files in the directory (i.e it seems to be enumerating all the files, but not properly excluding them)
10. quit uvicorn and restart it with `uvicorn --host 0.0.0.0 --port 4242 --reload --reload-exclude 'infra/**' myapp.api.main:app` (removing --reload-dir)
11. observe that the behaviour hasn't changed as far as I can tell - repeating step 6 will still trigger a reload

Output from uvicorn when I start it and modify the file in infra/packages/:
```
$ pwd 
/home/antisol/code/utest

$ uvicorn --host 0.0.0.0 --port 4242 --reload --reload-exclude 'infra/**' myapp.api.main:app
INFO:     Will watch for changes in these directories: ['/home/antisol/code/utest']
INFO:     Uvicorn running on http://0.0.0.0:4242 (Press CTRL+C to quit)
INFO:     Started reloader process [411700] using WatchFiles
INFO:     Started server process [411702]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
WARNING:  WatchFiles detected changes in 'infra/packages/foo/bar/baz.py'. Reloading...
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [411702]
INFO:     Started server process [411730]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```

Another note: providing --reload-dir but not --reload gives the following output:
`WARNING:  Current configuration will not reload as not all conditions are met, please refer to documentation.`

But I can't see anything about this in the documentation. I wouldn't expect to have to pass --reload if I'm passing --reload-dir, doesn't the one imply the other?

**Comment by schloerke** [comment]
When testing against a pattern, [exclusion is currently done](https://github.com/encode/uvicorn/blob/4fdfec4adf1ba6da5e65c8422321e203b6050052/uvicorn/supervisors/watchfilesreload.py#L48) by [`pathlib.Path`'s `.match(pattern)`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.match) method. This does not support `**` globs. It does support `*` globs. 😞 

I do not know of a way to ignore all folders that match a certain name: i.e. `__pycache__` or `.venv` folders that _could_ occur **anywhere** within the current working directory folder tree.

----------------------------------

The `FileFilter` class could be changed to not look for directories on initialization as directory structures can change over time.

The class below would allow for `--reload-exclude 'infra'`.  But it would also exclude `a/b/otherinfra/c/d.txt`. 

```python
class FileFilter:
    def __init__(self, config: Config):
        default_includes = ["*.py"]
        self.includes = [default for default in default_includes if default not in config.reload_excludes]
        self.includes.extend(config.reload_includes)
        self.includes = list(set(self.includes))

        default_excludes = [".*", ".py[cod]", ".sw.*", "~*"]
        self.excludes = [default for default in default_excludes if default not in config.reload_includes]
        self.excludes.extend(config.reload_excludes)
        self.excludes = list(set(self.excludes))

    def __call__(self, path: Path) -> bool:
        path_parts = [Path(part) for part in path.parts]

        for include_pattern in self.includes:
            if path.match(include_pattern):
                if str(path).endswith(include_pattern):
                    return True  # pragma: full coverage

                for exclude_pattern in self.excludes:
                    # Exclude if the pattern matches the file path directly
                    if path.match(exclude_pattern):
                        return False  # pragma: full coverage

                    # Exclude if the pattern matches any folder name directly
                    # (and doesn't contain a `/`)
                    if exclude_pattern.find("/") == -1 and any(part.match(exclude_pattern) for part in path_parts):
                        return False


                return True
        return False
```

----------------------

It would be best if `pathlib.Path`'s `.full_match(pattern)` method could be utilized. It supports `**` and would give expected behavior. However, it was introduced in python 3.13 😒 

[Using `glob.glob()`](https://docs.python.org/3/library/glob.html) would support `**`, but it actually tests the file system. While possibly expected behavior, it would be a massive slowdown in execution speed as it touches disk.

[Using `fnmatch.fnmatch(name, pat)`](https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch) would work well, but the pattern matching is not consistent with behavior we'd expect. Ex: `fnmatch.fnmatch("a/b/c.txt", "a/*.txt") # True`.

If `--reload-exclude-dir` was added, then exact matches to any directory name would be utilized. Ex: `str(part) == excluded_dir`. This is done within the [`watchfiles` `BaseFilter` class](https://github.com/samuelcolvin/watchfiles/blob/47f20c13debf8b0aabbdb9b86bb4182078f3a384/watchfiles/filters.py#L39-L55) which isn't utilized by uvicorn. But given the `*` matching of include/exclude, it feels like a half baked answer.

**Reply by schloerke** [reply]
I believe we could exclude exact matches to ancestor directories. It seems like a safe option.

**Reply by AntiSol** [reply]
> It would be best if pathlib.Path's .full_match(pattern) method could be utilized. It supports ** and would give expected behavior. However, it was introduced in python 3.13 😒
> 
> [Using glob.glob()](https://docs.python.org/3/library/glob.html) would support **, but it actually tests the file system. While possibly expected behavior, it would be a massive slowdown in execution speed as it touches disk.

I looked into this a little bit more after I reported the issue and came to this conclusion, too. I had been looking for a match function that worked without touching the disk, looks like `full_match` is what I wanted.

I agree that using `glob.glob` is not a good solution because it touches the disk, which makes it slow.

I'd prefer consistency with how shells tend to do globbing, so I'd prefer not to use `fnmatch`.

Perhaps we could use `full_match` and add this as a feature that's only available if you're using python>3.13? i.e have something like the following in `FileFilter.__init__`:
```
if not hasattr(pathlib.Path.full_match) and any([a for a in self.excludes if '**' in str(a)]): 
    raise NotImplementedError("You need at least python 3.13 to use ** in globs")
```
